Mathlib/LinearAlgebra/Multilinear/Basic.lean

```diff
@@ -1194,6 +1211,330 @@ theorem map_sub [DecidableEq ι] (m : ∀ i, M₁ i) (i : ι) (x y : M₁ i) :
   rw [sub_eq_add_neg, sub_eq_add_neg, MultilinearMap.map_add, map_neg]
 #align multilinear_map.map_sub MultilinearMap.map_sub
 
+/-- This calculates the differences between the values of a multilinear map at
+two arguments that differ on a finset `s` of `ι` of cardinality `n`. It requires a
+linear order on `ι` in order to choose an order isomorphism between `s` and `Fin n`
+and expresses the result using this order isomorphism.-/
+lemma map_sub_finset [LinearOrder ι] (f : MultilinearMap R M₁ M₂) (a b v : (i : ι) → (M₁ i))
+    {s : Finset ι} (hs : Finset.card s = n) :
+    f (s.piecewise a v) - f (s.piecewise b v) = Finset.sum Finset.univ (fun (i : Fin n) =>
+    f (fun j => if h : j ∈ s then (if (s.orderIsoOfFin hs).symm ⟨j, h⟩ < i then a j
+    else if (s.orderIsoOfFin hs).symm ⟨j, h⟩ = i then a j - b j else b j) else v j)) := by
```

Comment:
I don't understand why this is not stated with simply `f a - f b` as the LHS, without mentioning a Finset `s` at all. When `a` only differ from `b` on a Finset `s`, all summands outside of `s` would vanish by linearity, and you get a sum over `s.card` terms, and I think you can still get estimates as easily.

Reply:
Okay, I now see that induction on the Finset `s` is probably the easiest way to prove such statements, and we also don't want to restrict ourselves to finite `ι`. I have been able to give a short proof to a very similar result that doesn't require an Equiv with `Fin n`. The version with two `piecewise` can be obtained by combining two copies of `map_sub_map` below.
```
lemma map_update [DecidableEq ι] (x : (i : ι) → M₁ i) (i : ι) (v : M₁ i)  :
    f (update x i v) = f x - f (update x i (x i - v)) := by
  rw [map_sub, update_eq_self, sub_sub_cancel]

open Finset in
lemma map_sub_map [LinearOrder ι] (a b : (i : ι) → M₁ i) (s : Finset ι) :
    f a - f (s.piecewise b a) =
    ∑ i in s, f (fun j ↦ if j ∈ s → j < i then a j else if i = j then a j - b j else b j) := by
  refine s.induction_on_min ?_ fun k s hk ih ↦ ?_
  · rw [Finset.piecewise_empty, sum_empty, sub_self]
  rw [Finset.piecewise_insert, map_update, ← sub_add, ih,
      add_comm, sum_insert (lt_irrefl _ <| hk k ·)]
  simp_rw [s.mem_insert]
  congr 1
  · congr; ext i; split_ifs with h₁ h₂
    · rw [update_noteq, Finset.piecewise_eq_of_not_mem]
      · exact fun h ↦ (hk i h).not_lt (h₁ <| .inr h)
      · exact fun h ↦ (h₁ <| .inl h).ne h
    · cases h₂
      rw [update_same, s.piecewise_eq_of_not_mem _ _ (lt_irrefl _ <| hk k ·)]
    · push_neg at h₁
      rw [update_noteq (Ne.symm h₂), s.piecewise_eq_of_mem _ _ (h₁.1.resolve_left <| Ne.symm h₂)]
  · apply sum_congr rfl; intro i hi; congr; ext j; congr 1; apply propext
    simp_rw [imp_iff_not_or, not_or]; apply or_congr_left'
    intro h; rw [and_iff_right]; rintro rfl; exact h (hk i hi)
```

Reply:
Sorry, not by combining two copies, but by substitution (piecewise twice is the same as piecewise once). Updated my code above.

Reply:
Thanks ! Let me test it with the rest of the project (this is only a technical lemma used in a larger proof that continuous multilinear maps are differentiable, so I need it to work with the rest of the proof), and I'll get back to you.

Reply:
Yup it works, it actually simplifies the proof of the later results. :-) I'm rewriting the file to integrate your changes, thanks !

Reply:
Made the changes and pushed.